src/compiler/types.ts

```diff
@@ -981,12 +989,26 @@ module ts {
 
     // Return code used by getEmitOutput function to indicate status of the function
     export enum EmitReturnStatus {
-        Succeeded = 0,                      // All outputs generated if requested (.js, .map, .d.ts), no errors reported
-        AllOutputGenerationSkipped = 1,     // No .js generated because of syntax errors, nothing generated
-        JSGeneratedWithSemanticErrors = 2,  // .js and .map generated with semantic errors
-        DeclarationGenerationSkipped = 3,   // .d.ts generation skipped because of semantic errors or declaration emitter specific errors; Output .js with semantic errors
-        EmitErrorsEncountered = 4,          // Emitter errors occurred during emitting process
-        CompilerOptionsErrors = 5,          // Errors occurred in parsing compiler options, nothing generated
+        // All outputs generated if requested (.js, .map, .d.ts), no errors reported
+        Succeeded = 0,
+
+        // No .js, .map or d.ts generated because of diagnostics and the presence of the 
+        // -noEmitOnError optoin.
+        DiagnosticsPresent_AllOutputsSkipped = 1,
+
+        // .js and .map generated.  However, diagnostics were generated as well.
+        // No .d.ts was requested or generated.
+        DiagnosticsPresent_JavaScriptGenerated = 2,
+
+        // .js, .map generated.  .d.ts was requested but was not generated due to the 
+        // presence of diagnostics.
+        DiagnosticsPresent_JavaScriptGenerated_DeclarationNotGenerated = 3,
+
+        // Emitter errors occurred during emitting process.
+        EmitErrorsEncountered = 4,
+
+        // Errors occurred in parsing compiler options, nothing generated
+        CompilerOptionsErrors = 5,

```

Comment:
@ahejlsberg  I agree.  That's why i opened https://github.com/Microsoft/TypeScript/issues/1901 a few days back to track this and get it cleaned up.  As it affects partners, i wanted to make sure it was ok with them to clean up first.  Mohamed and i have discussed it and believe we should fix this up.  I will take care of this as part of this CR.